activerecord/lib/active_record/errors.rb

```diff
@@ -123,6 +123,28 @@ class RecordNotUnique < WrappedDatabaseException
   class InvalidForeignKey < WrappedDatabaseException
   end
 
+  # Raised when a foreign key constraint cannot be added because the column type does not match the referenced column type.
+  class MismatchedForeignKey < WrappedDatabaseException
+    def initialize(message)
+      parts = message.scan(/`(\w+)`[ $)]/).flatten
```

Comment:
@matthewd 

Personally, I'd prefer just this:

```ruby
  class MismatchedForeignKey < WrappedDatabaseException
    def initialize(message=nil)
      msg = "There is a mismatch between the foreign key and primary key column types.\n"
      msg << "Verify that the foreign key column type and the primary key of the associated table match types.\n"
      msg << message if message
      super(msg)
    end
  end
```

It gives plenty of context when `message` is passed in.

Current implementation:

```
Column `old_car_id` on table `engines` has a type of `bigint(20)`.
This does not match column `id` on `old_car`, which has type `int(11)`.
To resolve this issue, change the type of the `old_car_id` column on `engines` to be :integer. (For example `t.integer old_car_id`).

Mysql2::Error: Cannot add foreign key constraint: ALTER TABLE `engines` ADD CONSTRAINT `fk_rails_9f49f34f36`
FOREIGN KEY (`old_car_id`)
  REFERENCES `old_car` (`id`)
```

Proposed:

```
There is a mismatch between the foreign key and primary key column types.
Verify that the foreign key column type and the primary key of the associated table match types.

Mysql2::Error: Cannot add foreign key constraint: ALTER TABLE `engines` ADD CONSTRAINT `fk_rails_9f49f34f36`
FOREIGN KEY (`old_car_id`)
  REFERENCES `old_car` (`id`)
```